File: python/gtsam/examples/CustomFactorExample.py
"""
GTSAM Copyright 2010-2019, Georgia Tech Research Corporation,
Atlanta, Georgia 30332-0415
All Rights Reserved

See LICENSE for the license information

CustomFactor demo that simulates a 1-D sensor fusion task.
Author: Fan Jiang, Frank Dellaert
"""

from functools import partial
from typing import List, Optional

import gtsam
import numpy as np

I = np.eye(1)


def simulate_car() -> List[float]:
    """Simulate a car for one second"""
    x0 = 0
    dt = 0.25  # 4 Hz, typical GPS
    v = 144 * 1000 / 3600  # 144 km/hour = 90mph, pretty fast
    x = [x0 + v * dt * i for i in range(5)]

    return x


def error_gps(measurement: np.ndarray, this: gtsam.CustomFactor,
              values: gtsam.Values,
              jacobians: Optional[List[np.ndarray]]) -> float:
    """GPS Factor error function
    :param measurement: GPS measurement, to be filled with `partial`
    :param this: gtsam.CustomFactor handle
    :param values: gtsam.Values
    :param jacobians: Optional list of Jacobians
    :return: the unwhitened error
    """
    key = this.keys()[0]
    estimate = values.atVector(key)
    error = estimate - measurement
    if jacobians is not None:
        jacobians[0] = I

    return error


def error_odom(measurement: np.ndarray, this: gtsam.CustomFactor,
               values: gtsam.Values,
               jacobians: Optional[List[np.ndarray]]) -> float:
    """Odometry Factor error function
    :param measurement: Odometry measurement, to be filled with `partial`
    :param this: gtsam.CustomFactor handle
    :param values: gtsam.Values
    :param jacobians: Optional list of Jacobians
    :return: the unwhitened error
    """
    key1 = this.keys()[0]
    key2 = this.keys()[1]
    pos1, pos2 = values.atVector(key1), values.atVector(key2)
    error = measurement - (pos1 - pos2)
    if jacobians is not None:
        jacobians[0] = I
        jacobians[1] = -I

    return error


def error_lm(measurement: np.ndarray, this: gtsam.CustomFactor,
             values: gtsam.Values,
             jacobians: Optional[List[np.ndarray]]) -> float:
    """Landmark Factor error function
    :param measurement: Landmark measurement, to be filled with `partial`
    :param this: gtsam.CustomFactor handle
    :param values: gtsam.Values
    :param jacobians: Optional list of Jacobians
    :return: the unwhitened error
    """
    key = this.keys()[0]
    pos = values.atVector(key)
    error = pos - measurement
    if jacobians is not None:
        jacobians[0] = I

    return error


def main():
    """Main runner."""

    x = simulate_car()
    print(f"Simulated car trajectory: {x}")

    add_noise = True  # set this to False to run with "perfect" measurements

    # GPS measurements
    sigma_gps = 3.0  # assume GPS is +/- 3m
    g = [
        x[k] + (np.random.normal(scale=sigma_gps) if add_noise else 0)
        for k in range(5)
    ]

    # Odometry measurements
    sigma_odo = 0.1  # assume Odometry is 10cm accurate at 4Hz
    o = [
        x[k + 1] - x[k] +
        (np.random.normal(scale=sigma_odo) if add_noise else 0)
        for k in range(4)
    ]

    # Landmark measurements:
    sigma_lm = 1  # assume landmark measurement is accurate up to 1m

    # Assume first landmark is at x=5, we measure it at time k=0
    lm_0 = 5.0
    z_0 = x[0] - lm_0 + (np.random.normal(scale=sigma_lm) if add_noise else 0)

    # Assume other landmark is at x=28, we measure it at time k=3
    lm_3 = 28.0
    z_3 = x[3] - lm_3 + (np.random.normal(scale=sigma_lm) if add_noise else 0)

    unknown = [gtsam.symbol('x', k) for k in range(5)]

    print("unknowns = ", list(map(gtsam.DefaultKeyFormatter, unknown)))

    # We now can use nonlinear factor graphs
    factor_graph = gtsam.NonlinearFactorGraph()

    # Add factors for GPS measurements
    gps_model = gtsam.noiseModel.Isotropic.Sigma(1, sigma_gps)

    # Add the GPS factors
    for k in range(5):
        gf = gtsam.CustomFactor(gps_model, [unknown[k]],
                                partial(error_gps, np.array([g[k]])))
        factor_graph.add(gf)

    # New Values container
    v = gtsam.Values()

    # Add initial estimates to the Values container
    for i in range(5):
        v.insert(unknown[i], np.array([0.0]))

    # Initialize optimizer
    params = gtsam.GaussNewtonParams()
    optimizer = gtsam.GaussNewtonOptimizer(factor_graph, v, params)

    # Optimize the factor graph
    result = optimizer.optimize()

    # calculate the error from ground truth
    error = np.array([(result.atVector(unknown[k]) - x[k])[0]
                      for k in range(5)])

    print("Result with only GPS")
    print(result, np.round(error, 2),
          f"\nJ(X)={0.5 * np.sum(np.square(error))}")

    # Adding odometry will improve things a lot
    odo_model = gtsam.noiseModel.Isotropic.Sigma(1, sigma_odo)

    for k in range(4):
        odof = gtsam.CustomFactor(odo_model, [unknown[k], unknown[k + 1]],
                                  partial(error_odom, np.array([o[k]])))
        factor_graph.add(odof)

    params = gtsam.GaussNewtonParams()
    optimizer = gtsam.GaussNewtonOptimizer(factor_graph, v, params)

    result = optimizer.optimize()

    error = np.array([(result.atVector(unknown[k]) - x[k])[0]
                      for k in range(5)])

    print("Result with GPS+Odometry")
    print(result, np.round(error, 2),
          f"\nJ(X)={0.5 * np.sum(np.square(error))}")

    # This is great, but GPS noise is still apparent, so now we add the two landmarks
    lm_model = gtsam.noiseModel.Isotropic.Sigma(1, sigma_lm)

    factor_graph.add(
        gtsam.CustomFactor(lm_model, [unknown[0]],
                           partial(error_lm, np.array([lm_0 + z_0]))))
    factor_graph.add(
        gtsam.CustomFactor(lm_model, [unknown[3]],
                           partial(error_lm, np.array([lm_3 + z_3]))))

    params = gtsam.GaussNewtonParams()
    optimizer = gtsam.GaussNewtonOptimizer(factor_graph, v, params)

    result = optimizer.optimize()

    error = np.array([(result.atVector(unknown[k]) - x[k])[0]
                      for k in range(5)])

    print("Result with GPS+Odometry+Landmark")
    print(result, np.round(error, 2),
          f"\nJ(X)={0.5 * np.sum(np.square(error))}")


if __name__ == "__main__":
    main()

File: matlab/gtsam_examples/CustomFactorExample.m
import gtsam.*;

X = simulate_car();
fprintf('Simulated car trajectory: [%s]\n', num2str(X));

add_noise = true;

% GPS measurements
sigma_gps = 3.0;
gps_noise_model = noiseModel.Isotropic.Sigma(1, sigma_gps);
gps_sampler = Sampler(gps_noise_model, 42);
gps_measurements = zeros(1, 5);
for k=1:5
    gps_measurements(k) = X(k) + (add_noise * gps_sampler.sample());
end

% Odometry measurements
sigma_odo = 0.1;
odo_noise_model = noiseModel.Isotropic.Sigma(1, sigma_odo);
odo_sampler = Sampler(odo_noise_model, 42);
odometry_measurements = zeros(1, 4);
X_diff = diff(X);
for k=1:4
    odometry_measurements(k) = X_diff(k) + (add_noise * odo_sampler.sample());
end

% Landmark measurements:
sigma_lm = 1;
lm_noise_model = noiseModel.Isotropic.Sigma(1, sigma_lm);
lm_sampler = Sampler(lm_noise_model, 42);

lm_0 = 5.0;
landmark0_measurement = X(1) - lm_0 + (add_noise * lm_sampler.sample());
lm_3 = 28.0;
landmark3_measurement = X(4) - lm_3 + (add_noise * lm_sampler.sample());

unknown = [];
for k=0:4
    unknown = [unknown symbol('x', k)];
end

% 1. Result with only GPS
factor_graph = NonlinearFactorGraph;
factors = {}; % Keep alive
for k=1:5
    error_func = make_gps_callback(gps_measurements(k));
    gps_factor = CustomFactor(gps_noise_model, unknown(k), error_func);
    factor_graph.add(gps_factor);
    factors{end+1} = gps_factor;
end

v = Values;
for i=1:5
    v.insert(unknown(i), [0.0]);
end

params = GaussNewtonParams;
result = GaussNewtonOptimizer(factor_graph, v, params).optimize();

vals = zeros(1, 5);
for i=1:5, vals(i) = result.atVector(unknown(i)); end
fprintf('Result with only GPS: [%s], Error: %f\n', num2str(round(vals, 2)), factor_graph.error(result));

% 2. Result with GPS+Odometry
for k=1:4
    error_func = make_odom_callback(odometry_measurements(k));
    odometry_factor = CustomFactor(odo_noise_model, [unknown(k), unknown(k + 1)], error_func);
    factor_graph.add(odometry_factor);
    factors{end+1} = odometry_factor;
end

result = GaussNewtonOptimizer(factor_graph, v, params).optimize();
vals = zeros(1, 5);
for i=1:5, vals(i) = result.atVector(unknown(i)); end
fprintf('Result with GPS+Odometry: [%s], Error: %f\n', num2str(round(vals, 2)), factor_graph.error(result));

% 3. Result with GPS+Odometry+Landmark
error_func0 = make_lm_callback(lm_0 + landmark0_measurement);
factor0 = CustomFactor(lm_noise_model, unknown(1), error_func0);
factor_graph.add(factor0);
factors{end+1} = factor0;

error_func3 = make_lm_callback(lm_3 + landmark3_measurement);
factor3 = CustomFactor(lm_noise_model, unknown(4), error_func3);
factor_graph.add(factor3);
factors{end+1} = factor3;

result = GaussNewtonOptimizer(factor_graph, v, params).optimize();
vals = zeros(1, 5);
for i=1:5, vals(i) = result.atVector(unknown(i)); end
fprintf('Result with GPS+Odometry+Landmark: [%s], Error: %f\n', num2str(round(vals, 2)), factor_graph.error(result));

%% Callback Factory Helpers
function f = make_gps_callback(measurement)
    f = @(varargin) error_gps(measurement, varargin{:});
end

function f = make_odom_callback(measurement)
    f = @(varargin) error_odom(measurement, varargin{:});
end

function f = make_lm_callback(measurement)
    f = @(varargin) error_lm(measurement, varargin{:});
end

%% Helper Functions
function x = simulate_car()
    x0 = 0;
    dt = 0.25;
    v = 144 * 1000 / 3600;
    x = x0 + v * dt * (0:4);
end

function varargout = error_gps(measurement, this, values)
    keys = this.keys();
    estimate = values.atVector(keys.at(0));
    residual = estimate - measurement;
    if nargout > 1
        varargout{1} = residual;
        varargout{2} = {1};
    else
        varargout{1} = residual;
    end
end

function varargout = error_odom(measurement, this, values)
    keys = this.keys();
    pos1 = values.atVector(keys.at(0));
    pos2 = values.atVector(keys.at(1));
    residual = (pos2 - pos1) - measurement;
    if nargout > 1
        varargout{1} = residual;
        varargout{2} = {-1, 1};
    else
        varargout{1} = residual;
    end
end

function varargout = error_lm(measurement, this, values)
    keys = this.keys();
    pos = values.atVector(keys.at(0));
    residual = pos - measurement;
    if nargout > 1
        varargout{1} = residual;
        varargout{2} = {1}; % Jacobian w.r.t. x
    else
        varargout{1} = residual;
    end
end
